### PR TITLE
MI-2218: Icon Component

### PR DESCRIPTION
**Pull request by tanisha-agarwal**
### Ticket Link
 [MI-2218](https://brightscout.atlassian.net/browse/MI-2218)

### Description
- Created Icon Component
- Updated Esbuild config 
- Tested Build for Icon Component

### Screenshots
image
image

### CODE REVIEW

File: esbuild.dev.config.ts [hunk 1]
```diff
@@ -16,16 +15,16 @@ esbuild
...
       sourcemap: true,
-      external: [...Object.keys(dependencies), 'react/jsx-runtime'],
+      external: ['react/jsx-runtime'],
       plugins: [
+        svgrPlugin({ exportType: 'named' }),
         inlineImage(),
         sassPlugin({
           async transform(source) {
-            const { css } = await postcss([autoprefixer]).process(source);
+            const { css } = await postcss([autoprefixer]).process(source, { from: undefined });
```
Comment (raghavaggarwal2308) [inline esbuild.dev.config.ts:24]:
```suggestion
            const {css} = await postcss([autoprefixer]).process(source, {from: undefined});
```
We don't add padding in curly braces, right?.....fix it everywhere

**Review by raghavaggarwal2308 — APPROVED** [review]

File: src/components/Icon/Icon.component.tsx [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+ * ```ts
+ * <Icon name="Unlink" />
+ * ```
+ *
+ * @example Using custom size and color
+ * ```ts
+ * <Icon name="Unlink" size="36" iconColor="#000" />
+ * ```
+ */
+export const Icon = (props: IconProps) => {
```
Comment (ayusht2810) [inline src/components/Icon/Icon.component.tsx:20]: can we destructure the props directly here?

**Review by ayusht2810 — APPROVED** [review]

File: src/components/Icon/Icon.component.tsx (hunk 2 above)
Comment (tanisha-agarwal) [reply src/components/Icon/Icon.component.tsx:20]: We can, but I think we should not expose props in the functional component's argument. I have destructured props directly in the function's argument for sub-components.

File: esbuild.dev.config.ts (hunk 1 above)
Comment (avas27JTG) [reply esbuild.dev.config.ts]: wondering why the linter didn't give error for this? @tanisha-agarwal

File: package.json [hunk 3]
```diff
@@ -65,10 +68,11 @@
     "postcss": "~8.4.16",
     "postcss-cli": "~10.0.0",
     "prettier": "~2.7.1",
-    "react": "~16.13.1",
+    "react": "~16.14.0",
```
Comment (avas27JTG) [inline package.json:71]: make sure this react version is the same as that of Mattermost v6.0.0

File: package.json [hunk 4]
```diff
@@ -65,10 +68,11 @@
...
     "react-bootstrap": "~2.5.0",
```
Comment (avas27JTG) [inline package.json:72]: check for all major dependencies like react, redux, bootstrap, etc used here matched with MM v6.0.0, if not then we will need to upgrade/downgrade to match MM

File: src/components/Icon/Icon.component.tsx [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+ * Icon Component
+ *
+ * @example Correct usage with default size and color
+ * ```ts
+ * <Icon name="Unlink" />
+ * ```
+ *
+ * @example Using custom size and color
+ * ```ts
+ * <Icon name="Unlink" size="36" iconColor="#000" />
```
Comment (avas27JTG) [inline src/components/Icon/Icon.component.tsx:17]: nice way of explaining through the comments

File: src/components/Icon/Icon.component.tsx [hunk 6]
```diff
@@ -0,0 +1,36 @@
+import React, {createElement} from 'react';
+
+import {IconProps} from './Icon';
```
Comment (avas27JTG) [inline src/components/Icon/Icon.component.tsx:3]: why are we keeping file names with a capital letter?

File: src/components/Icon/Icon.d.ts [hunk 7]
```diff
@@ -0,0 +1,41 @@
+import IconMap from './Icon.map';
+
+export type IconType = keyof typeof IconMap;
+
+export interface StyledContainerProps {
```
Comment (avas27JTG) [inline src/components/Icon/Icon.d.ts:5]: no need to export as the file type is `.d.ts` so, any types mentioned here would automatically be available in all `ts` files without explicit export/import

File: src/components/Icon/Icon.component.tsx (hunk 6 above)
Comment (tanisha-agarwal) [reply src/components/Icon/Icon.component.tsx:3]: In Frontend, we usually keep `Components` files and folders name's with the first letter as capital

The reason for using the first capital letter is that when you search for a file by name in the project you can always recognize components just by file name `header` vs `Header`, you always know that the second one is a component and the first one is either container or page.

File: src/components/Icon/Icon.d.ts (hunk 7 above)
Comment (avas27JTG) [reply src/components/Icon/Icon.d.ts:5]: okay, didn't know about this issue with storyBook

**Review by avas27JTG — APPROVED** [review]

**Pull request merged**